경북대 BE 김진현 - Step 1

kimjhyun0627 opened the pull request:
*늦어서 죄송합니다 ㅜㅜ*

# 구현한 기능

- config 리팩토링
- JPA 의존성 추가
- entity 리팩토링
- 레포지토리 리팩토링 → Service 레이어까지 리팩토링
- 단위 테스트 일부 수정 & JpaTest로 레포지토리 테스트 코드 작성

kimjhyun0627 committed:
docs: README 업데이트

kimjhyun0627 committed:
feat: JPA 의존성 추가

kimjhyun0627 committed:
refactor: config 리팩토링

kimjhyun0627 committed:
refactor: entity 리팩토링

kimjhyun0627 committed:
refactor: repository 리팩토링

kimjhyun0627 committed:
feat: 테스트 코드 작성

Gyuchool commented on README.md:
```diff
@@ -1,14 +1,25 @@
+
 ## STEP 1
 
+* [X] config 리팩토링
```
기능 구현 목록 정리👍

Gyuchool commented on src/main/java/gift/controller/user/WishController.java:
```diff
@@ -30,30 +30,30 @@ public WishController(WishService wishService) {
-        Wish w = wishService.addWish(member.getId().id(), request.productId(), request.amount());
-        return WishResponse.of(w.getId(), w.getProductId(), w.getAmount());
+        Wish w = wishService.addWish(member, request.productId(), request.amount());
+        return WishResponse.of(w.getId(), w.getProduct().getId(), w.getAmount());
```
w만 보면 어떤 것을 뜻하는지 알기 어려운것 같습니다! 적절한 변수명을 지어주세요🙏

Gyuchool commented on src/main/java/gift/entity/member/Member.java:
```diff
@@ -3,21 +3,47 @@
-    private final Role role;
-    private final LocalDateTime createdAt;
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
```
GenerateTyoe을 IDENTITY로 설정하신 이유가 있을까요? 여러 방법이 있을텐데 INDENTITY를 선택하신 이유가 궁금합니다!

kimjhyun0627 replied:
이전까지 JPA를 사용하면 항상 `IDENTITY`를 사용했어서 습관적으로 사용한 것 같습니다.(IntelliJ 자동 완성 역시 이렇게 작동해서 더 아무 생각 없이 짠 것 같기도 하네요...)
조금 더 찾아보니 지금은 간단한 CRUD 위주라 `IDENTITY`를 사용해도 문제가 발생하지 않는 것 같지만, DB 이식성을 고려해서 `AUTO`를 쓰는 방법이나 추후 대량 배치가 발생한다면 `SEQUENCE`를 사용하는 방법도 있을 것 같습니다!

Gyuchool replied:
IDENTITY를 적용했을때 db 벤더마다 어떻게 기본키 값이 증가되는지 확인해도 좋겠네요:)

kimjhyun0627 replied:
값 건너뛰기가 발생할 수 있군요...! `SEQUENCE`를 사용하자니 시퀀스를 생성해야 하고...DB는 참 어려운 것 같습니다,,,

kimjhyun0627 replied:
SEQUENCE를 생성하고 SEQUENCE를 사용할까 하다가, 문득 과제에서 제공한 스키마가 생각나서 일단은 IDENTITY를 유지한 채로 PR 보내드립니다..!
_방금 전까지 스켈레톤 코드라고 판단해 임의로 제약조건 추가하고는 갑자기 그 스키마를 따라간다고 말씀드리기 민망하긴 하네요...:)_

Gyuchool replied:
넵 현재 IDENTITY 전략 좋은 것 같습니다👍

Gyuchool commented on src/main/java/gift/entity/member/Member.java:
```diff
@@ -3,21 +3,47 @@
+    @Embedded
+    private MemberPasswordHash passwordHash;
+
+    @Convert(converter = RoleConverter.class)
```
Converter를 따로 만드신 이유가 있으실까요?

kimjhyun0627 replied:
코드리뷰 요청 PR 당시에는 `enum` 타입을 어떻게 DB에 안정적으로 저장할 수 있을지 찾아보다가 발견한 방법이었는데, 조금은 복잡한 것 같아 이후에 추가 학습했습니다.
`@Enumerated`를 사용해 Converter 없이도 `enum`을 사용할 수 있게 변경했습니다.

Gyuchool commented on src/main/java/gift/entity/member/Member.java:
```diff
@@ -3,21 +3,47 @@
+    @Column(name = "id")
+    private Long id;
+
+    @Embedded
+    private MemberEmail email;
+
+    @Embedded
+    private MemberPasswordHash passwordHash;
```
각각의 객체가 책임을 가지고있군요👍 유지보수하기 좋아보이네요!
이렇게 객체로 감싸서 사용하면서 느꼈던 장단점도 공유해주시면 좋을 것 같습니다!

kimjhyun0627 replied:
이전 미션 멘토님께서 `value object`로 감싸는 현재 방식을 추천해주셨는데, 도메인 필드 각각에 대한 제약조건을 관리하고 변경하는 것이 훨씬 직관적이라고 생각해 사용하고 있습니다.
다만, JPA를 사용하며 DB의 `AUTO_INCREMENT`를 사용해 값을 생성하는 각 도메인의 `ID`값과의 충돌이 발생해서, 현재는 `ID`값만 `Long` 타입으로 받고 있습니다!

kimjhyun0627 replied:
존재하던 각 필드의 `ID` object의 경우, deprecate하기는 싫어서 서비스 레이어에서 값 검증을 할 때 사용할 수 있도록 `getId()`에서 사용해 값 검증을 할 수 있게 해두었습니다.

kimjhyun0627 replied:
`productPrice`나 `wishAmount`와 같이 비교적 제약조건이 단순하고 변경이 적은 필드에 대해서도 `value object`를 사용해야 하는지는 아직은 잘 와닿지 않지만, 코드의 통일성을 위해 제작해두었던 것 같습니다. 해당 두 필드는 엔티티 레벨의 제약조건보다는 서비스 레이어에서 처리할 법한 비즈니스 로직이 더 많을 것 같은데, 그렇다면 이들 필드의 래핑은 개발 단계에서의 선택이 되는 걸까요?

Gyuchool replied:
진현님이 모든 필드를 vo로 감싸면서 느끼셨던 부분을 기억해두면 좋을 것 같아요! 어디까지 감쌀지는 정답이 없어요:) 

저도 vo를 감쌀때 지금처럼 모든 필드를 감싸면 응집도는 높아지지만 매번 래핑해야하고 코드 추적할때 어려움이 따라왔던 경험이 있어요. 앞으로 진현님도 이때의 경험을 기반으로 어디까지 vo로 감쌀지 기준을 잡으시면 좋을 것 같습니다👍

Gyuchool commented on src/main/java/gift/entity/member/Member.java:
```diff
@@ -3,21 +3,47 @@
 
     private Member(
-            MemberId id,
+            Long id,                       // ← MemberId 대신 Long
```
주석은 제거하는게 좋아보이는데 어떠신가요?🙄

Gyuchool commented on src/main/java/gift/entity/member/value/MemberEmail.java:
```diff
@@ -1,10 +1,16 @@
+        String email
+) {
 
     public static final String EMAIL_REGEX =
```
접근제어자를 바꿔도 좋겠네요!

kimjhyun0627 replied:
현재 value object가 각각 value 패키지 내부에 있는데, 접근제어자가 public이 아니라면 접근이 불가능할 것 같습니다..!

Gyuchool replied:
`private static final String EMAIL_REGEX =`
이 부분은 패키지와 관련이 없지 않나요?🙄

Gyuchool commented on src/main/java/gift/entity/member/Member.java:
```diff
@@ -3,21 +3,47 @@
+    @Column(name = "created_at", nullable = false, updatable = false)
+    private LocalDateTime createdAt;
+
+    protected Member() {
```
기본생성자는 언제 쓰이기때문에 protected로 선언하셨을까요?

kimjhyun0627 replied:
JPA가 엔티티 인스턴스를 내부적으로 생성할 때 `public` 혹은 `protected` 접근제어자를 가진 인자 없는(파라미터 없는) 생성자가 필요하고 해서, 상대적으로 private한 `protected` 접근제어자를 사용했습니다 !

Gyuchool replied:
혹시 왜 필요 없는지도 알고 계실까요?

Gyuchool commented on src/main/java/gift/entity/member/value/Role.java:
```diff
@@ -1,9 +1,24 @@
     USER,
     ADMIN;
 
+    public static Role of(String input) {
```
정적 팩토리 메서드 네이밍을 하실때 어떤 이유로 of를 쓰신걸까요?

kimjhyun0627 replied:
예전에 스프링을 잘 다루는 친구가 단순한 역할을 하는 메소드일수록 `of()`처럼 간결한 스타일의...? 함수명을 잘 사용하는 것도 중요하다고 말해줘서, 메소드를 만들 때 `valueOf()`와 `of()` 사이에서 고민 중 그 친구에서 다시 물어봤었습니다..! 레퍼런스는 <effiective java>라고 하는데, 이 책은 아직 읽어보지 못했지만 비슷한 책인 <clean code>에서도 비슷한 내용이 있었던 것 같아 적용해 보았습니다 !

Gyuchool replied:
넵 아마 이펙티브 자바를 말씀하신 것 같은데 책에서 추천해주는 네이밍을 적용하신 것 같아서 질문드렸습니다:)
이펙티브 자바에서 of()는 여러 매개변수를 받을 때 사용하는걸 추천해줘서요:)

kimjhyun0627 replied:
어라...코멘트에서 < 와 > 사이에 넣으니 없어져버리네요...!
친구가 추천해 준 책은 "이펙티브 자바", 제가 훑어본 책은 "클린 코드" 였습니다 !

Gyuchool commented on src/main/java/gift/entity/wish/Wish.java:
```diff
@@ -1,96 +1,118 @@
+    }
+
+    @Override
+    public boolean equals(Object o) {
```
id를 기준으로 equals&hascode를 재정의하신 이유가 있을까요?
equals&hashcode를 재정의한 이유가 무엇인지 궁금합니다:)

kimjhyun0627 replied:
`Wish` 클래스의 `equals`를 내부에서 사용하므로 재정의했던 것 같은데, 이 과정에서 IntelliJ의 `equals() 및 hashCode() 생성` 기능을 모든 엔티티에 대해서 남용한게 아닌가 생각이 듭니다... `Wish`의 `equals`를 제외한 불필요한 오버라이드 삭제하겠습니다 !

Gyuchool commented on src/main/java/gift/entity/wish/Wish.java:
```diff
@@ -1,96 +1,118 @@
+@Entity
+@Table(
+        name = "wish",
+        uniqueConstraints = @UniqueConstraint(columnNames = {"member_id", "product_id"})
```
<img width="556" height="392" alt="스크린샷 2025-07-16 오전 4 51 36" src="https://github.com/user-attachments/assets/c9fd9724-2a52-449b-bf65-fce717c5c442" />

요구사항에는 unique제약 조건이 없는 것 같습니다! 추가하신 이유가 있으실까요?

kimjhyun0627 replied:
이 부분은 제가 과제 요구사항을 위시리스트보다는 장바구니로 인식해서 그렇게 했던 것 같습니다 ! 
사실 과제 요구사항이 위시리스트이든 장바구니이든, 동일한 사용자가 같은 상품을 중복해서 담는다는 것이 올바르지 않다고 생각했고, 제공받은 SQL 스키마 역시 이전 요구사항에 의거, 상품 리스트의 숨김 처리를 위해 hidden 속성을 추가함으로 인해 이미 깨져버린 상황에서 참고를 위한 스켈레톤 코드라고 생각해 엔티티에서 제약조건을 제거하기보다는 스키마에 제약조건을 추가하는 방향으로 설계했습니다.

Gyuchool commented on src/main/java/gift/repository/member/AdminAccountInitializer.java:
```diff
@@ -25,59 +27,47 @@ public AdminAccountInitializer(MemberRepository memberRepository) {
+        );
     }
 
     @EventListener(ContextRefreshedEvent.class)
```
초기화 방식이 여러가지 있을 것 같은데 `@EventListener(ContextRefreshedEvent.class)` 를 사용한 이유가 있을까요?

kimjhyun0627 replied:
이 부분은 구현 과정에서 여러 블로그와 LLM의 도움을 받았는데, 당시에 `@PostConstruct`가 빈 초기화 순서를 보장해주지 못한다는 블로그 글을 읽었던 것 같아 비교적 안전한 방법이라고 판단해 `@EventListener(ContextRefreshedEvent.class)`를 사용했습니다!

Gyuchool commented on src/main/java/gift/repository/wish/WishRepository.java:
```diff
@@ -1,18 +1,21 @@
+public interface WishRepository extends JpaRepository<Wish, Long> {
 
-    List<Wish> findByMember(MemberId memberId);
+    @EntityGraph(attributePaths = {"member", "product"})
```
EntitiyGraph는 무엇인가요? 사용해보면서 어떤 장단점으로 선택하신걸까요?

kimjhyun0627 replied:
JPQL보다 간결하게 선언 가능했고, 위시리스트는 개인이 관리하므로 트래픽이 심각하게 과도하게 발생하지는 않을 것이라 생각했습니다.
추가적으로, 구현해두고 실제로 사용하지 않던 메소드는 삭제했습니다 !

Gyuchool commented on src/main/java/gift/service/member/MemberServiceImpl.java:
```diff
@@ -26,46 +27,30 @@ public MemberServiceImpl(MemberRepository memberRepository, JwtUtil jwtUtil) {
-        m = memberRepository.register(m);
-        String token = jwtUtil.generateToken(m.getId().id(), m.getRole());
+        Member member = Member.register(authRequest.email(), hash);
+        try {
+            member = memberRepository.save(member);
+        } catch (DataIntegrityViolationException e) {
+            throw new MemberAlreadyExistsException(authRequest.email());
+        }
```
try-catch로 감싼 목적이 무엇일까요?

kimjhyun0627 replied:
register에서 사용하는 `save()` 의 경우, 동일 이메일로 인한 DB의 UNIQUE 제약 위반이 발생하면 `DataIntegrityViolationException`을 던지는데, 이를 catch하여 custom exception으로 매핑하기 위해서입니다!

Gyuchool commented on src/main/java/gift/service/member/MemberServiceImpl.java:
```diff
@@ -26,46 +27,30 @@ public MemberServiceImpl(MemberRepository memberRepository, JwtUtil jwtUtil) {
     @Override
     public AuthResponse login(String email, String rawPassword) {
-        Member m = memberRepository.findByEmail(email)
+        Member member = memberRepository.findByEmail_Email(email)
```
변수명이 명확해졌군요👍

Gyuchool commented on src/main/java/gift/service/wish/WishServiceImpl.java:
```diff
@@ -34,45 +36,52 @@ public WishServiceImpl(
+    public Wish updateWish(Long wishId, Member member, Long productId, int amount) {
+        Wish existing = wishRepository.findById(wishId)
+                .orElseThrow(() -> new WishNotFoundException(wishId));
+        if (!existing.isOwnedBy(member)) {
```
existing이란 변수명보다 명확한 변수명이 더 좋을 것 같습니다!

Gyuchool commented on src/main/java/gift/service/wish/WishServiceImpl.java:
```diff
@@ -34,45 +36,52 @@ public WishServiceImpl(
         }
-        existing.withAmount(amount);
-        return wishRepository.update(existing);
+        return wishRepository.save(existing.withAmount(amount));
```
메서드 이름은 updateWish()인데 save()를 하니 insert하는 것처럼 느껴지는데 어떻게 개선할 수 있을까요?

kimjhyun0627 replied:
현재 JPA 엔티티를 최대한 불변하게 유지하고자 `setter`를 사용하지 않고 있는데, 엔티티를 불변으로 유지하지 않는다면 `@Transcation`과 `setter`를 사용해 
```
existingWish.setAmount(amount));
```
이처럼 dirty checking을 사용해 명시적으로 수정할 수 있을 것 같습니다.
다만, 아직까지는 엔티티를 mutable하게 변경하고 싶지는 않아 코멘트를 보더라도 따로 변경하지는 않았습니다.
이 부분은 어떤 방법이 낫다라고 조언해주실 수 있으시다면 빠르게 반영해보겠습니다..!

kimjhyun0627 replied:
메소드 이름만 `changeWishAmount()`로 변경했습니다 !

Gyuchool replied:
의도적으로 dirtyChecking을 사용하지 않으셨군요👍
아시다시피 두 방법 모두 사용이 가능해서요. 어떤게 낫다는 없긴하지만 한 가지 궁금한 점이 떠올라서 질문드릴게욤:)

진현님은 JPA를 왜 사용하시는지 고민하신 적이 있으실까요? jdbc와 jpa사용하면서 어떤 차이점을 느끼셨을까요?

kimjhyun0627 replied:
일단 엔티티와 DB 사이의 매핑이 명확하게 된다는 점이 제일 크게 와닿은 것 같습니다. 레포지토리를 interface로 자체적으로 구현해, SQL문을 최소한으로 작성해도 된다는 점 역시 중요했던 것 같습니다.
조금 더 찾아보니, JPA는 '엔티티가 곧 스키마다'라는 느낌이 강한 것 같습니다..!
엔티티의 메소드를 추가하고, 각 서비스 레이어의 업데이트 관련 부분만 우선적으로 dirty checking 방식으로 변경했습니다 !

kimjhyun0627 replied:
다른 메소드들도 dirty check가 가능한지 학습해보겠습니다 !
→ 리뷰 요청 후 update: 권장되지 않는군요...!

Gyuchool replied:
좋습니다👍 제가 질문을 jpa에 집중해서 드렸군요.! JPA보단 ORM이 왜 등장하였는지 이번 미션을 통해 체감해보시죠!

Gyuchool commented on src/test/java/gift/repository/member/MemberRepositoryTest.java:
```diff
@@ -0,0 +1,83 @@
+import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
+import org.springframework.dao.DataIntegrityViolationException;
+
+@DataJpaTest
```
dataJpaTest, jdbcTest, SpringBootTest 3개의 테스트는 각각 언제 사용하는 것일까요?

kimjhyun0627 replied:
`@DataJpaTest`는 JPA 리포지토리와 엔티티 매핑을 검증할 때,
`@JdbcTest`는 JdbcTemplate 기반의 순수 SQL·DAO 레이어 테스트,
`@SpringBootTest`의 경우에는 애플리케이션 전체 컨텍스트를 띄워 컨트롤러부터 서비스, 리포지토리까지 통합 검증할 때 사용한다고 알고 있습니다!

Gyuchool replied:
잘 정리해주셨군요👍

Gyuchool reviewed (changes requested):
안녕하세요 진현님
이전 미션에서 꼼꼼한 테스트와 validation 그리고 객체에게 책임을 할당하신 모습들 너무 인상적이네요💯💯
whish테이블에서 unique설정 관련해서 확인 요청 드립니다!

몇 가지 코멘트를 남겨두었으니 반영 및 의견 주시고 다시 요청주세요!🙏
궁금한 점은 디스코드 DM이나 PR 커멘트로 언제든지 여쭤봐주세요!(새벽도 가능!)

kimjhyun0627 committed:
refactor: step 1 코드리뷰 반영

- schema.sql 수정: 한 회원이 두개의 같은 상품 등록 방지
- 각 객체에 맞는 변수명 리팩토링
- RoleConverter 삭제
- `updateWish()` -> `changeWishAmount()` 메소드명 변경

kimjhyun0627 commented:
`whish테이블에서 unique설정 관련해서 확인 요청 드립니다!`
 → 관련 스키마 수정했습니다 !

Gyuchool reviewed (changes requested):
안녕하세요 진현님
빠르게 피드백 반영하고 요청주셨네요💯

다만, 중간에 리뷰에 대한 답변이 누락된 것 같아서 해당 부분 DM으로 전달드렸는데 확인 한 번 부탁드릴게요!
확인해주시고 다시 요청주시면 병합시키겠습니다😊

kimjhyun0627 committed:
refactor: step 1 코드리뷰 2차 반영

- entity 수정: 변경 가능한 항목에 대한 change 함수 생성 & 불필요한 오버라이드 제거
- repository 수정: 불필요한 메소드 제거
- 스키마 수정: 표준 문법으로 수정
- service 수정: entity 수정에 따른 update 관련 함수들 로직 변경(dirty check)
- 위 사항들에 해당하는 테스트 코드 변경

kimjhyun0627 committed:
fix: MemberServiceImpl 트랜잭션 실행

- MemberServiceImpl.java에 @Transactional 어노테이션 추가

kimjhyun0627 commented:
코드리뷰 요청드린 직후 `MemberServiceImpl` 클래스에 `@Transactional` 어노테이션 누락을 확인해 커밋 하나가 추가되었습니다ㅜㅜ
즉시 fix commit 올렸습니다 ! 참고 부탁드리겠습니다 !!

Gyuchool reviewed (approved):
안녕하세요 진현님!
빠르게 피드백 반영해주셨군요👍
이번 단계는 여기서 병합하겠습니다! 다음 단계에서 뵈어요👏
